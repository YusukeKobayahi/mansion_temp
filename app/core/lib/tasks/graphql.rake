# frozen_string_literal: true

module GraphqlTask
  extend Rake::DSL

  namespace :graphql do
    namespace :consumer do
      task schema_dump: :environment do
        require 'graphql/rake_task'

        GraphQL::RakeTask.new(
          load_schema: lambda { |_task|
            require File.expand_path('../../app/graphql/consumer/inno_schema', __dir__)
            Consumer::InnoSchema
          },
          directory: './graphql/consumer',
        )
        Rake::Task['graphql:schema:idl'].invoke
      end
    end
  end
end
