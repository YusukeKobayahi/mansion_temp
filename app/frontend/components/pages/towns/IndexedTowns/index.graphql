fragment MsLibraryCitiesIndexedTowns on City {
  name
  prefecture {
    name
  }
  indexedTowns {
    id
    name
    mansionCount
  }
}
