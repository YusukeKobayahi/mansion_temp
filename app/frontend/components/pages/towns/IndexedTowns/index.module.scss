@import "~/styles/variables";
@import "~/styles/utils";

.container {
  margin-top: 50px;
}

.list {
  display: flex;
  flex-wrap: wrap;
  margin: 0 12px;

  @include screenMd {
    padding: 0;
    margin: -20px 0 0;
  }
}
