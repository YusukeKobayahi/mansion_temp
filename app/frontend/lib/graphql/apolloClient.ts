// 参考
// https://github.com/vercel/next.js/tree/canary/examples/with-apollo

import { useMemo } from "react";
import { ApolloClient, ApolloLink, InMemoryCache } from "@apollo/client";
import { createHttpLink } from "apollo-link-http";
import { from } from "apollo-link";
import { createFragmentArgumentLink } from "apollo-link-fragment-argument";
import { mergeDeep, relayStylePagination } from "@apollo/client/utilities";
import { isSSR } from "~/lib/utils";

/* eslint-disable @typescript-eslint/no-explicit-any -- キャッシュの型を事前に決められないため */
type CacheState = Record<string, any> | null;
/* eslint-enable @typescript-eslint/no-explicit-any */

/* eslint-disable @typescript-eslint/naming-convention -- GraphQLのタイプ名 (このconventionに従わない)を指定したいので */
const createCache = () =>
  new InMemoryCache({
    typePolicies: {
      Mansion: {
        keyFields: ["uniqueCode"],
        fields: {
          salesHistories: relayStylePagination(),
        },
      },
    },
  });
/* eslint-enable @typescript-eslint/naming-convention */

const link = from([
  createFragmentArgumentLink(),
  createHttpLink({
    uri: `${process.env.NEXT_PUBLIC_CORE_API_URL}/consumer/graphql`,
    fetchOptions: {
      mode: "cors",
    },
  }),
]);

function createApolloClient() {
  return new ApolloClient({
    ssrMode: isSSR(),
    // See: https://github.com/apollographql/apollo-link/issues/1258#issuecomment-604031488
    // TODO: @apollo/clientバージョン上げたら直るかも
    link: (link as unknown) as ApolloLink,
    cache: createCache(),
    defaultOptions: {
      watchQuery: {
        // SSR時にuseQueryによってリクエストしてレンダリングする場合、
        // getServerSidePropsからキャッシュが渡っている必要がある。(でないと描画されない)
        // そのため、SSR時は常にcacheから取得されるべきで、実装意図を明確にするためPolicyも設定しておく
        // (エラーになるわけではない)
        fetchPolicy: isSSR() ? "cache-only" : "cache-first",
      },
    },
  });
}

let apolloClient: ReturnType<typeof createApolloClient>;

export function initializeApollo(
  initialState: CacheState = null
): ApolloClient<CacheState> {
  const _apolloClient = apolloClient ?? createApolloClient();

  // If your page has Next.js data fetching methods that use Apollo Client, the initial state
  // gets hydrated here
  if (initialState) {
    // Get existing cache, loaded during client side data fetching
    const existingCache = _apolloClient.extract();

    // Merge the existing cache into data passed from getStaticProps/getServerSideProps
    const data = mergeDeep(initialState, existingCache);

    // Restore the cache with the merged data
    _apolloClient.cache.restore(data);
  }
  // For SSG and SSR always create a new Apollo Client
  if (isSSR()) return _apolloClient;
  // Create the Apollo Client once in the client
  if (!apolloClient) apolloClient = _apolloClient;

  return _apolloClient;
}

export function useApollo(initialState: CacheState): ApolloClient<CacheState> {
  const store = useMemo(() => initializeApollo(initialState), [initialState]);
  return store;
}
