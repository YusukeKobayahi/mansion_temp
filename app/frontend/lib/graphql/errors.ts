export interface GraphQLError {
  graphQLErrors?: {
    message: string;
    extensions?: {
      code: string;
    };
  }[];
}

export const isNotFoundError = (err: GraphQLError): boolean => {
  const graphQLErrors = err.graphQLErrors;
  if (graphQLErrors == null) return false;

  return graphQLErrors[0]?.extensions?.code === "NOT_FOUND";
};
