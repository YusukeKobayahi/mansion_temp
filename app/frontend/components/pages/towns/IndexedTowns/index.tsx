import { MsLibraryCitiesIndexedTownsFragment } from "~/graphql/generated";

import styles from "./index.module.scss";
import BoxHeading from "~/components/commons/headings/BoxHeading";
import LinkBlock from "~/components/commons/LinkBlock";

export interface IndexedTownsProps {
  data: MsLibraryCitiesIndexedTownsFragment;
}

const IndexedTowns: React.FC<IndexedTownsProps> = ({
  data,
}: IndexedTownsProps) => {
  if (!data.indexedTowns || data.indexedTowns.length === 0) return null;

  const items = data.indexedTowns.map((town) => {
    if (town.mansionCount > 0) {
      return (
        <LinkBlock
          key={town.id}
          name={town.name}
          id={town.id}
          mansionCount={town.mansionCount}
          link="towns"
          eventParams={{
            action: "msl_area_click",
            category: "msl_ichiran",
            label: "msl_indexed_towns",
            value: 1,
          }}
        />
      );
    }
  });

  return (
    <section className={styles.container}>
      <h2>
        <BoxHeading>
          {data.prefecture.name}
          {data.name}の人気のエリアから中古マンションを探す
        </BoxHeading>
      </h2>
      <ul className={styles.list}>{items}</ul>
    </section>
  );
};

export default IndexedTowns;
