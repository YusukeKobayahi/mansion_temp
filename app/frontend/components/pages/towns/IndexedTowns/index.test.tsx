import { render } from "@testing-library/react";
import mockData from "~/components/pages/towns/fixtures/mockData.json";
import { TownsQuery } from "~/graphql/generated";
import IndexedTowns from "~/components/pages/towns/IndexedTowns";

describe("components/pages/cities/IndexedTowns", () => {
  it("正常に表示される", () => {
    const mock = mockData as TownsQuery;
    const { asFragment } = render(<IndexedTowns data={mock.towns[0].city} />);
    expect(asFragment()).toMatchSnapshot();
  });
});
