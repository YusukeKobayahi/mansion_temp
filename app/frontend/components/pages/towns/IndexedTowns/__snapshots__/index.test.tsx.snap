// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components/pages/cities/IndexedTowns 正常に表示される 1`] = `
<DocumentFragment>
  <section
    class="container"
  >
    <h2>
      <div
        class="heading"
      >
        東京都目黒区の人気のエリアから中古マンションを探す
      </div>
    </h2>
    <ul
      class="list"
    >
      <li
        class="item"
      >
        <button
          class="link"
          href="/towns/39025"
        >
          青葉台
          <span
            class="count"
          >
            (7件)
          </span>
        </button>
      </li>
      <li
        class="item"
      >
        <button
          class="link"
          href="/towns/39028"
        >
          柿の木坂
          <span
            class="count"
          >
            (5件)
          </span>
        </button>
      </li>
      <li
        class="item"
      >
        <button
          class="link"
          href="/towns/39031"
        >
          駒場
          <span
            class="count"
          >
            (5件)
          </span>
        </button>
      </li>
      <li
        class="item"
      >
        <button
          class="link"
          href="/towns/39044"
        >
          碑文谷
          <span
            class="count"
          >
            (14件)
          </span>
        </button>
      </li>
      <li
        class="item"
      >
        <button
          class="link"
          href="/towns/39049"
        >
          目黒本町
          <span
            class="count"
          >
            (16件)
          </span>
        </button>
      </li>
      <li
        class="item"
      >
        <button
          class="link"
          href="/towns/39050"
        >
          八雲
          <span
            class="count"
          >
            (8件)
          </span>
        </button>
      </li>
    </ul>
  </section>
</DocumentFragment>
`;
